File: agnt-plugin/reports.js
// PulseMesh JS — Reports & Baselines & History
// Port of Python reports.py + baselines.py + history.js

import fs from 'fs';
import path from 'path';
import { nowISO, loadJSON, writeJSON } from './util.js';

// ---- Reports ----

export function generateMarkdownReport(summary) {
  const mesh = summary.mesh || {};
  const sensors = summary.sensors || [];
  const alerts = summary.alerts || [];
  const lines = [
    `# PulseMesh Report`,
    `**Generated:** ${summary.timestamp || nowISO()}`,
    `**Run ID:** ${summary.run_id || 'unknown'}`,
    ``,
    `## Mesh Summary`,
    `| Metric | Value |`,
    `|---|---|`,
    `| Mesh Health | ${(mesh.mesh_health ?? 0).toFixed(3)} |`,
    `| Sensors | ${mesh.sensor_count ?? 0} |`,
    `| Fallback | ${mesh.fallback_sensor_count ?? 0} |`,
    `| Alerts | ${mesh.alert_count ?? 0} |`,
    `| Critical | ${mesh.critical_alert_count ?? 0} |`,
    ``,
    `## Sensors`,
    `| Sensor | Provider | Health | Anomaly | Coherence | Live |`,
    `|---|---|---|---|---|---|`,
  ];
  for (const s of sensors) {
    const m = s.metrics || {};
    lines.push(`| ${s.label || s.profile_id} | ${s.provider} | ${(m.health_score ?? 0).toFixed(3)} | ${(m.anomaly_score ?? 0).toFixed(3)} | ${(m.coherence_avg ?? 0).toFixed(3)} | ${s.used_live_data ? '✓' : '✗'} |`);
  }
  if (alerts.length > 0) {
    lines.push(``, `## Alerts`, `| Severity | Sensor | Message |`, `|---|---|---|`);
    for (const a of alerts) {
      lines.push(`| ${a.severity} | ${a.profile_id} | ${a.message || ''} |`);
    }
  }
  return lines.join('\n');
}

export function compareSummaries(beforePath, afterPath) {
  const before = typeof beforePath === 'string' ? loadJSON(beforePath) : beforePath;
  const after = typeof afterPath === 'string' ? loadJSON(afterPath) : afterPath;
  if (!before || !after) throw new Error('Could not load summaries');
  const bSensors = Object.fromEntries((before.sensors || []).map(s => [s.profile_id, s]));
  const aSensors = Object.fromEntries((after.sensors || []).map(s => [s.profile_id, s]));
  const allIds = [...new Set([...Object.keys(bSensors), ...Object.keys(aSensors)])].sort();
  const sensors = allIds.map(id => {
    const old = bSensors[id];
    const neu = aSensors[id];
    if (!old) return { profile_id: id, status: 'added' };
    if (!neu) return { profile_id: id, status: 'removed' };
    const om = old.metrics || {};
    const nm = neu.metrics || {};
    return {
      profile_id: id, status: 'changed',
      health_delta: (nm.health_score ?? 0) - (om.health_score ?? 0),
      anomaly_delta: (nm.anomaly_score ?? 0) - (om.anomaly_score ?? 0),
      coherence_delta: (nm.coherence_avg ?? 0) - (om.coherence_avg ?? 0),
      live_data_changed: !!old.used_live_data !== !!neu.used_live_data,
    };
  });
  return {
    before_run: before.run_id, after_run: after.run_id,
    before_timestamp: before.timestamp, after_timestamp: after.timestamp,
    sensors,
  };
}

// ---- Baselines ----

const TRACKED_METRICS = ['mean', 'coherence_avg', 'health_score', 'anomaly_score', 'volatility', 'drift', 'stability_fraction'];

export function loadBaselines(baselinePath) {
  if (!fs.existsSync(baselinePath)) return { version: '0.2', updated_at: null, sensors: {} };
  try {
    const obj = loadJSON(baselinePath);
    if (!obj || typeof obj !== 'object') return { version: '0.2', updated_at: null, sensors: {} };
    obj.version = '0.2';
    obj.sensors = obj.sensors || {};
    return obj;
  } catch {
    return { version: '0.2', updated_at: null, sensors: {} };
  }
}

export function updateBaselines(summaryPath, baselinePath, window = 50) {
  const summary = typeof summaryPath === 'string' ? loadJSON(summaryPath) : summaryPath;
  const baselines = loadBaselines(baselinePath);
  const sensors = baselines.sensors || {};

  for (const sensor of (summary.sensors || [])) {
    const sid = sensor.profile_id;
    if (!sid) continue;
    const entry = sensors[sid] = sensors[sid] || { profile_id: sid, label: sensor.label, provider: sensor.provider, samples: [] };
    const sample = {
      timestamp: summary.timestamp || nowISO(),
      run_id: summary.run_id,
      used_live_data: !!sensor.used_live_data,
      metrics: Object.fromEntries(TRACKED_METRICS.filter(k => k in (sensor.metrics || {})).map(k => [k, sensor.metrics[k]])),
    };
    entry.samples.push(sample);
    if (entry.samples.length > window) entry.samples = entry.samples.slice(-window);
  }

  baselines.updated_at = nowISO();
  writeJSON(baselinePath, baselines);
  return baselines;
}

export function annotateWithBaselines(summary, baselinePath) {
  const baselines = loadBaselines(baselinePath);
  for (const sensor of (summary.sensors || [])) {
    const entry = baselines.sensors?.[sensor.profile_id];
    if (!entry || entry.samples.length < 2) continue;
    const prev = entry.samples[entry.samples.length - 2];
    const cur = sensor.metrics || {};
    const baseline = {};
    for (const k of TRACKED_METRICS) {
      if (k in cur && k in prev.metrics) {
        baseline[k] = {
          current: cur[k],
          previous: prev.metrics[k],
          delta: cur[k] - prev.metrics[k],
          z_score: entry.samples.length >= 3 ? computeZ(entry.samples.map(s => s.metrics[k]), cur[k]) : 0,
        };
      }
    }
    sensor.baseline = baseline;
  }
  return summary;
}

function computeZ(history, current) {
  const n = history.length;
  if (n < 3) return 0;
  const mean = history.reduce((a, b) => a + b, 0) / n;
  const variance = history.reduce((a, b) => a + (b - mean) ** 2, 0) / n;
  const std = Math.sqrt(variance);
  return std > 1e-9 ? (current - mean) / std : 0;
}

// ---- History ----

export function loadLedger(ledgerPath) {
  if (!fs.existsSync(ledgerPath)) return [];
  return fs.readFileSync(ledgerPath, 'utf-8').split('\n').filter(Boolean).map(line => {
    try { return JSON.parse(line); } catch { return null; }
  }).filter(Boolean);
}

export function summarizeHistory(runsDir, limit = 20) {
  const ledger = loadLedger(path.join(runsDir, 'ledger.jsonl'));
  const recent = ledger.slice(-limit);
  const health = recent.map(r => r.mesh?.mesh_health ?? 0);
  const trend = health.length >= 2 ? health[health.length - 1] - health[0] : 0;
  return {
    run_count: recent.length,
    health_trend: trend,
    latest_health: health[health.length - 1] ?? 0,
    runs: recent.map(r => ({
      run_id: r.run_id,
      timestamp: r.timestamp,
      mesh_health: r.mesh?.mesh_health ?? 0,
      alert_count: r.mesh?.alert_count ?? 0,
      fallback_count: r.mesh?.fallback_sensor_count ?? 0,
    })),
  };
}

File: src/pulsemesh/reports.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .util import load_json, now_iso, write_json


def load_summary(path: Path) -> dict[str, Any]:
    obj = load_json(path)
    if not isinstance(obj, dict) or "mesh" not in obj or "sensors" not in obj:
        raise ValueError(f"not a PulseMesh summary: {path}")
    return obj


def compare_summaries(before_path: Path, after_path: Path) -> dict[str, Any]:
    before = load_summary(before_path)
    after = load_summary(after_path)
    before_sensors = {s["profile_id"]: s for s in before.get("sensors", [])}
    after_sensors = {s["profile_id"]: s for s in after.get("sensors", [])}
    all_ids = sorted(set(before_sensors) | set(after_sensors))
    sensors = []
    for sid in all_ids:
        old = before_sensors.get(sid)
        new = after_sensors.get(sid)
        if old is None:
            sensors.append({"profile_id": sid, "status": "added"})
            continue
        if new is None:
            sensors.append({"profile_id": sid, "status": "removed"})
            continue
        old_m = old.get("metrics", {})
        new_m = new.get("metrics", {})
        sensors.append({
            "profile_id": sid,
            "status": "changed",
            "health_delta": float(new_m.get("health_score", 0.0)) - float(old_m.get("health_score", 0.0)),
            "anomaly_delta": float(new_m.get("anomaly_score", 0.0)) - float(old_m.get("anomaly_score", 0.0)),
            "coherence_delta": float(new_m.get("coherence_avg", 0.0)) - float(old_m.get("coherence_avg", 0.0)),
            "live_data_changed": bool(old.get("used_live_data")) != bool(new.get("used_live_data")),
        })
    return {
        "timestamp": now_iso(),
        "before": str(before_path),
        "after": str(after_path),
        "mesh": {
            "health_delta": float(after["mesh"].get("mesh_health", 0.0)) - float(before["mesh"].get("mesh_health", 0.0)),
            "coherence_delta": float(after["mesh"].get("mesh_coherence", 0.0)) - float(before["mesh"].get("mesh_coherence", 0.0)),
            "fallback_delta": int(after["mesh"].get("fallback_sensor_count", 0)) - int(before["mesh"].get("fallback_sensor_count", 0)),
        },
        "sensors": sensors,
    }


def write_markdown_report(summary_path: Path, out_path: Path) -> None:
    summary = load_summary(summary_path)
    mesh = summary.get("mesh", {})
    lines = [
        f"# PulseMesh Report: {summary.get('run_id')}",
        "",
        f"- Generated: {now_iso()}",
        f"- Summary: `{summary_path}`",
        f"- Sensors: {mesh.get('sensor_count', 0)}",
        f"- Live sensors: {mesh.get('live_sensor_count', 0)}",
        f"- Fallback sensors: {mesh.get('fallback_sensor_count', 0)}",
        f"- Mesh health: {float(mesh.get('mesh_health', 0.0)):.4f}",
        f"- Mesh coherence: {float(mesh.get('mesh_coherence', 0.0)):.4f}",
        "",
        "## Sensors",
        "",
        "| Sensor | Provider | Live | Health | Anomaly | Stability | Drift |",
        "|---|---:|---:|---:|---:|---:|---:|",
    ]
    for sensor in summary.get("sensors", []):
        metrics = sensor.get("metrics", {})
        lines.append(
            "| {label} | {provider} | {live} | {health:.4f} | {anomaly:.4f} | {stable:.4f} | {drift:.4f} |".format(
                label=str(sensor.get("label", sensor.get("profile_id", ""))).replace("|", "\\|"),
                provider=sensor.get("provider", ""),
                live="yes" if sensor.get("used_live_data") else "no",
                health=float(metrics.get("health_score", 0.0)),
                anomaly=float(metrics.get("anomaly_score", 0.0)),
                stable=float(metrics.get("stability_fraction", 0.0)),
                drift=float(metrics.get("drift", 0.0)),
            )
        )
    lines.append("")
    alerts = summary.get("alerts", [])
    if alerts:
        lines.extend(["## Alerts", ""])
        for alert in alerts:
            lines.append(f"- **{alert.get('severity', 'warning')}** `{alert.get('profile_id')}`: {alert.get('message')} (value={alert.get('value')})")
        lines.append("")
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text("\n".join(lines), encoding="utf-8")


def write_compare_json(before_path: Path, after_path: Path, out_path: Path) -> None:
    write_json(out_path, compare_summaries(before_path, after_path))

